py4j-python/src/py4j/tests/java_dir_test.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import unicode_literals, absolute_import
+
+
+from py4j.java_gateway import JavaGateway, GatewayParameters, java_import,\
+    UserHelpAutoCompletion
+from py4j.tests.java_gateway_test import (
+    start_example_app_process)
+from contextlib import contextmanager
+from nose.tools import eq_
+
+ExampleClassFields = sorted([
+                      "field10",
+                      "field11",
+                      "field20",
+                      "field21",
+                      "static_field"
+                      ])
+
+ExampleClassMethods = sorted([
+                      # From ExampleClass
+                      "method1",
+                      "method2",
+                      "method3",
+                      "method4",
+                      "method5",
+                      "method6",
+                      "method7", # overloaded
+                      "method8",
+                      "method9",
+                      "method10", # overloaded
+                      "method11",
+                      "getList",
+                      "getField1",
+                      "setField1",
+                      "getStringArray",
+                      "getIntArray",
+                      "callHello",
+                      "callHello2",
+                      "static_method",
+
+                      # From Object
+                      "getClass",
+                      "hashCode",
+                      "equals",
+                      "toString",
+                      "notify",
+                      "notifyAll",
+                      "wait"
+                      ])
+
+ExampleClassStatics = sorted([
+                              "StaticClass",
+                              "static_field",
+                              "static_method"
+                              ])
+
+@contextmanager
+def example_app_process():
+    p = start_example_app_process()
+    try:
+        yield p
+    finally:
+        p.join()
+
+@contextmanager
+def gateway(*args, **kwargs):
+    g = JavaGateway(
+        gateway_parameters=GatewayParameters(*args, auto_convert=True, **kwargs))
+    lineSep = g.jvm.System.lineSeparator()
+    try:
+        yield g
+        # Call a dummy method to make sure we haven't corrupted the streams
+        eq_(lineSep, g.jvm.System.lineSeparator())
```

Comment:
> why did you need to do that? (I'm curious, nothing against it)
Because it was possible to write to the stream extraneous characters or insufficient characters. I wrote this after I realized I was testing a bunch of stuff, but I hadn't put the final separator at the end of the command.